### PR TITLE
Add an off drawer canvas region for slide-in menu bar

### PR DESCRIPTION
**Pull request by alice-0-kim**
Initial pull request to `develop` branch.

### CODE REVIEW

**Comment by alice-0-kim** [comment]
Hi Joel, no worries. Thank you for the time you've spent on the code review. Please let me know if you have additional comments.

Thank you!

File: template.php [hunk 1]
```diff
@@ -479,7 +519,7 @@ function megatron_megatron_btn_dropdown($variables) {
 
   // Finish markup
   $output .= '
-  <div class="ubc7-arrow down-arrow"></div>
+  <div class="ubc7-arrow right-arrow"></div>
```
Comment (joelpittet) [inline template.php]: These may break BC, could we use maybe a different theme?

File: template.php [hunk 2]
```diff
@@ -919,7 +959,7 @@ function megatron_megatron_links($variables) {
         if (count($children) > 0) {
           $link['html'] = TRUE;
           $output .=  '<div class="btn-group">' .l($link['title'], $link['href'], $link);
-          $output .=  '<button class="btn dropdown-toggle" data-toggle="dropdown"><span class="ubc7-arrow blue down-arrow"></span></button>';
+          $output .=  '<button class="btn dropdown-toggle" data-toggle="dropdown"><span class="ubc7-arrow white right-arrow"></span></button>';
```
Comment (joelpittet) [inline template.php]: Same here with blue and white change could have some negative changes to existing sites?

File: template.php (hunk 1 above)
Comment (alice-0-kim) [reply template.php]: Yes, it should be down-arrow. Thank you for bringing this up.

**Comment by alice-0-kim** [comment]
@joelpittet Thank you for bringing up this issue. I've committed the changes so that the arrow on the main navigation menu a blue arrow pointing down while the arrow on the drawer region is a white arrow pointing to the right.

File: css/off.canvas.drawer.css [hunk 3]
```diff
@@ -0,0 +1,241 @@
+#drawer-button {
+  float:right;
+  color:#000;
+  border:none;
+  background-color:rgba(0, 0, 0, 0);
+  font-family: Arial,Sans-Serif;
+  padding:10px;
+  font-size:14px;
+  line-height: 1.75em!important;
```
Comment (joelpittet) [inline css/off.canvas.drawer.css]: It may be good if we don't have to use `!important`, It's an id so it's pretty strong in specificity and harder to override in a child theme.

File: css/off.canvas.drawer.css [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+  width: 32px;
+}
+
+.drawer-toggle--hamburger,.drawer-toggle--hamburger:hover {
+  box-shadow: 1px 1px 1px #191938;
+  -moz-box-shadow: 1px 1px 1px #191938;
+}
+
+.drawer-toggle--hamburger:hover {
+  -webkit-box-shadow: inset 1px 1px 1px #191938;
```
Comment (joelpittet) [inline css/off.canvas.drawer.css]: Should this be in the above statement on line 28? It seems like it would only work in webkit on hover.

File: css/off.canvas.drawer.css [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+  box-shadow: 2px 0 2px rgba(0, 0, 0, .15);
+}
+#pushed-content.drawer--push-right.off-canvas-drawer--is-active {
+  transform: translate3D(20rem, 0, 0);
+  box-shadow: 2px 0 2px rgba(0, 0, 0, .15);
+}
+#off-canvas-drawer--primary {
+  transition: transform 0.125s ease, visibility 0.125s ease;
+  visibility: visible;
+  height: 100% !important;
```
Comment (joelpittet) [inline css/off.canvas.drawer.css]: !important needed? May be good to look at all these?

File: templates/page/page.tpl.php [hunk 6]
```diff
@@ -132,6 +138,13 @@
   <nav id="ubc7-unit-menu" role="navigation" class="navbar expand">
     <div class="navbar-inner expand">
       <?php print $fluidcontainerstart; ?>
+        <?php if ($drawer_enabled): ?>
+          <!-- UBC Unit Drawer Menu Button -->
+          <button style="float:right;margin-right:17px;" class="hamburger visible-md visible-lg drawer-toggle--primary" id="drawer-button">
+            <span>Menu ☰</span>
```
Comment (joelpittet) [inline templates/page/page.tpl.php]: I like the hamburger icon but maybe it should be consistent with the other hamburger with `<span class="icon-bar"></span>`?

File: templates/page/page.tpl.php [hunk 7]
```diff
@@ -111,7 +117,7 @@
     <?php print $fluidcontainerstart; ?>
     <div class="span12">
       <div class="navbar">
-        <a class="btn btn-navbar" data-toggle="collapse" data-target="#ubc7-unit-navigation">
+        <a class="btn btn-navbar drawer-toggle--primary" data-toggle="collapse" data-target="#ubc7-unit-navigation">
```
Comment (joelpittet) [inline templates/page/page.tpl.php]: Is this class needed since it's below?

File: templates/page/page.tpl.php [hunk 8]
```diff
@@ -132,6 +138,13 @@
   <nav id="ubc7-unit-menu" role="navigation" class="navbar expand">
     <div class="navbar-inner expand">
       <?php print $fluidcontainerstart; ?>
+        <?php if ($drawer_enabled): ?>
+          <!-- UBC Unit Drawer Menu Button -->
+          <button style="float:right;margin-right:17px;" class="hamburger visible-md visible-lg drawer-toggle--primary" id="drawer-button">
```
Comment (joelpittet) [inline templates/page/page.tpl.php]: The inline styles could be moved into the CSS file?

**Comment by alice-0-kim** [comment]
@joelpittet I removed most of the important keywords and moved in-line styles to a CSS file. For the hamburger icon, did you want to keep the 'Menu' text besides the icon? Please let me know if further changes are required.

Thank you!

**Comment by joelpittet** [comment]
The 'Menu' text is fine with me, maybe a question for @occupant but I'd say it could use a span and BEM class around the text to make it easier to hide/remove.

Thanks for fixing the styles @alice-0-kim 

Any thoughts on the last item? `drawer-toggle--primary` added in two places, is that needed?

**Comment by alice-0-kim** [comment]
@joelpittet No worries at all. I'm assuming that you're referring to line 120 and 143 on the most updated `page.tpl.php`. I tried removing them, but it no longer triggers the action once `drawer-toggle--primary` gets removed. However, I might be looking at the wrong direction so please point me to the right direction.

File: template.php [hunk 9]
```diff
@@ -334,8 +334,44 @@ function megatron_preprocess_page(&$variables) {
...
+  }
+
+  // Add js and css for drawer option
+  if ($drawer_enabled) {
+    drupal_add_js(drupal_get_path('theme', 'megatron') . '/js/off.canvas.drawer.js');
+    drupal_add_css(drupal_get_path('theme', 'megatron') . '/css/off.canvas.drawer.css');
+  }
+
+  // Add js and css for navigation sticky option
+  if (theme_get_setting('clf_sticky_option')) {
```
Comment (joelpittet) [inline template.php]: Maybe this sticky option can be a separate pull request or is it needed for this to work?

File: js/off.canvas.drawer.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+      pushPanel.addEventListener("transitionend", finishedAnimating);
+      function finishedAnimating() {
+        document.body.classList.remove(animatingClass);
+        pushPanel.classList.remove(animatingClass);
+        pushContent.classList.remove(animatingClass);
+        pushMask.classList.remove(animatingClass);
+      }
+    }
+  }
+}(jQuery));
```
Comment (joelpittet) [inline js/off.canvas.drawer.js]: Closures are good but you aren't using jQuery, maybe that can just not pass params & args in.

File: theme-settings.php [hunk 11]
```diff
@@ -150,6 +136,64 @@ function megatron_form_system_theme_settings_alter(&$form, &$form_state) {
...
+    '#prefix' => '<div class="clf_navigation_option">',
+    '#suffix' => '</div>',
+    '#collapsible' => TRUE,
+    '#collapsed' => TRUE,
+    '#weight' => -7,
+  );
+
+  $form['clf_navigation_option']['clf_navigation_placement'] = array(
+    '#type' => 'select',
+    '#title' => t('Choose the type of primary navigation that should be used on this website.'),
```
Comment (joelpittet) [inline theme-settings.php]: This may be better to have as a separate option related to drawer only because the primary can be left there and have a drawer with the region there now. hmmm, thoughts?

File: templates/page/page.tpl.php [hunk 12]
```diff
@@ -132,6 +138,18 @@
   <nav id="ubc7-unit-menu" role="navigation" class="navbar expand">
     <div class="navbar-inner expand">
       <?php print $fluidcontainerstart; ?>
+        <?php if ($drawer_enabled): ?>
+          <!-- UBC Unit Drawer Menu Button -->
+          <button class="hamburger visible-md visible-lg drawer-toggle--primary" id="drawer-button">
+            <div>Menu</div>
```
Comment (joelpittet) [inline templates/page/page.tpl.php]: Span tag would be better here and the text could be nice to have it configurable with that as the default in the appearance settings.

File: templates/page/page.tpl.php (hunk 7 above)
Comment (joelpittet) [reply templates/page/page.tpl.php]: I'll try this out

**Comment by alice-0-kim** [comment]
@joelpittet I also think it is appropriate to have a separate pull request for adding sticky option. Also, should I wait to see if James has any input on the last comment you made about having a separate option related to drawer?

Thank you!

**Comment by joelpittet** [comment]
@occupant may have better wording and ideas on the names for the options. My guess would be it may be nice to start with "Use drawer menu?" And that could show additional options like naming the drawer menu title, the animation style, etc.  The one thing we agreed on for sure was the ability to name the drawer menu toggle in the appearance options.

File: js/off.canvas.drawer.js [hunk 13]
```diff
@@ -0,0 +1,37 @@
+(function () {
+
+  Drupal.behaviors.megatronDrawer = {
+    attach: function(context, settings) {
+      // Primary drawer toggle.
+      var pushTrigger = document.getElementsByClassName("drawer-toggle--primary");
```
Comment (joelpittet) [inline js/off.canvas.drawer.js]: This won't work in IE8, I'm fine with that but may want to double check with James to see what they are supporting. 
https://caniuse.com/#search=getElementsByClassName

Comment (occupant) [reply js/off.canvas.drawer.js]: I don't think that should be an issue - we don't support anything older than IE10 for new projects (unless someone wants to make that a requirement, in which case we would expect to make adjustments). I figure IE9 is a very reasonable baseline these days.

Comment (occupant) [reply js/off.canvas.drawer.js]: I should have prefaced that with "Of course, this is just my opinion".

File: css/off.canvas.drawer.css [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+.off-canvas-drawer .ubc7-arrow {
+  background-position: -1113px -227px !important;
+}
+#drawer-button .icon-bar {
+  display:block;
+  width:18px;
+  height:2px;
+  background-color:#000;
+  -webkit-border-radius:1px;
+  -moz-border-radius:1px;
```
[the remark this answers is not in the document]
Comment (occupant) [reply css/off.canvas.drawer.css]: Assuming IE9 is our baseline, we should also remove box-shadow prefixes 
https://caniuse.com/#search=box-shadow

[the remark this answers is not in the document]
Comment (occupant) [reply css/off.canvas.drawer.css]: Also, it looks like the less source file isn't committed - @alice-0-kim would it be possible to add that?

File: theme-settings.php [hunk 15]
```diff
@@ -150,6 +136,57 @@ function megatron_form_system_theme_settings_alter(&$form, &$form_state) {
...
+    '#default_value' => theme_get_setting('clf_navigation_placement'),
+    '#options' => array(
+      'default' => t('Default CLF - Horizontal'),
+      'drawer--push-left' => t('Left Push Drawer'),
```
Comment (occupant) [inline theme-settings.php]: Maybe to make the labels clearer 

```
Default: no drawer region
Drawer Region: push from left
Drawer Region: cover from left
Drawer Region: push from right
Drawer Region: cover from right
```

And I guess the `clf_navigation_placement` setting is poorly named, since it's not just for navigation. Maybe something more like `clf_drawer_region`?

File: css/off.canvas.drawer.css (hunk 14 above)
[the remark this answers is not in the document]
Comment (alice-0-kim) [reply css/off.canvas.drawer.css]: @occupant perhaps you are referring to the `less` folder? If not, please point me in the right direction.

[the remark this answers is not in the document]
Comment (occupant) [reply css/off.canvas.drawer.css]: @alice-0-kim yes, that's right, the .less files in the less directory.

[the remark this answers is not in the document]
Comment (alice-0-kim) [reply css/off.canvas.drawer.css]: @occupant I see! I just included `off.canvas.drawer.less` file to `less` folder, and it still needs a review on naming variables.

File: js/off.canvas.drawer.js (hunk 13 above)
Comment (joelpittet) [reply js/off.canvas.drawer.js]: Ok cool, this review can be closed.

File: css/off.canvas.drawer.css (hunk 14 above)
[the remark this answers is not in the document]
Comment (joelpittet) [reply css/off.canvas.drawer.css]: The autoprefixer pull request can deal with this review in a more generic way:) Thanks again for the PR @alice-0-kim
